huggingface_pipelines/pipeline.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+from typing import List, TypedDict, Dict, Any
+import logging
+from dataclasses import dataclass, replace
+from datasets import Dataset
+import os
+from contextlib import contextmanager
+import torch
+from .dataset import DatasetConfig
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class PipelineOverwrites(TypedDict):
+    batch_size: int
+    device: str
+    cache_to_arrow: bool
+    take: int
+    output_dir: str
+    output_file_name: str
+    columns: List[str]
+
+
+@dataclass
+class PipelineConfig(ABC):
+    """
+    Abstract base class for pipeline configurations.
+
+    Attributes:
+        dataset_name (str): The name of the HuggingFace dataset to be used.
+        dataset_split (str): The dataset split to be used (e.g., 'train', 'test', 'validation').
+        batch_size (int): The batch size to be used for processing.
+        device (str): The device to use for inference (e.g., 'cpu', 'cuda').
+        cache_to_arrow (bool): Whether to cache results to Arrow format. Defaults to False.
+        output_dir (str): The directory to save the output to. Defaults to 'results'.
+        take (int): The number of batches to take for processing. Defaults to -1 (process all).
+        dataset_uuid (str): The id for the dataset instance, this is used for caching. Defaults to None.
+
+    """
+    columns: List[str]
+    dataset_config: DatasetConfig
+    batch_size: int = 5
+    device: str = "cpu"
+    take: int = -1
+    cache_to_arrow: bool = False
+
+    def with_overwrites(self, overwrites: PipelineOverwrites):
+        return replace(self, **overwrites)
+
+
+class Pipeline(ABC):
+    """
+    Abstract base class for different pipelines.
+    """
+
+    def __init__(self, config: PipelineConfig):
+        self.config = config
+        self.results = []
```

Comment:
why having this attribute ? For sure we dont want to store the resulting data attached to the class I already comment this.